### PR TITLE
EZP-26476: Wider roles being ignored if you define more restrictive roles

### PR DESCRIPTION
**Pull request by alongosz**
> Status: **Ready for a review**
> Fixes: [EZP-26476](https://jira.ez.no/browse/EZP-26476)

If a role with wide policy (e.g. module `content` for all functions with no limitations) contains also a more specific policy (e.g. module `content` for the `read` function with some limitations) and a role is assigned to a group limited by a Subtree, then wider role is ignored. Adding any limitation to a group assignment is the root cause of this issue. Relevant code that is not working properly is in [this scope](https://github.com/ezsystems/ezpublish-kernel/blob/v6.5.0/eZ/Publish/Core/Repository/Repository.php#L414-L451).

**TODO**:
- [x] Create integration unit test case reproducing EZP-26476 use case (41c3ef2).
- [x] Actual fix: Remove narrow policies when wider ones w/o any limitations exist (237f926).
- [x] ~Adjust tests in `RepositoryTest` class to account for changes.~

### CODE REVIEW

**Comment by andrerom** [comment]
Could you also add a separate test where the wider policy has function defined? like:
```php
$this->addPolicyToNewRole($roleCreateStruct, 'content', 'read', []);
```

**Comment by alongosz** [comment]
> Could you also add a test where the wider policy has function defined? like:
> ```php
> $this->addPolicyToNewRole($roleCreateStruct, 'content', 'read', []);
> ```

It failed, so thanks :-)

> And one where user has full access:
> ```php
> $this->addPolicyToNewRole($roleCreateStruct, '*', '*', []);
> ```

Both tests added in 4af6ece using proper `@dataProvider`.

I've added better (maybe) fix for this in 5f760e0 (will be described in diff comment).

I'm open to any ideas for more extensive tests.

File: eZ/Publish/Core/Repository/Values/User/Policy.php [hunk 1]
```diff
@@ -33,4 +33,19 @@ public function getLimitations()
...
+     * @param \eZ\Publish\Core\Repository\Values\User\Policy $policy
+     * @return bool
+     */
+    public function isNarrow(Policy $policy)
+    {
+        return
+            !empty($this->limitations) && empty($policy->getLimitations()) &&
+            ($this->module === $policy->module || $policy->module === '*') &&
+            ($this->function === $policy->function || $policy->function === '*')
+        ;
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Values/User/Policy.php]: To simplify [this code](https://github.com/alongosz/ezpublish-kernel/blob/d912295e74e37ecbeb318b7d9f6953076836f243/eZ/Publish/Core/Repository/Repository.php#L481), I've decided to move checking if policy is more narrow to `\eZ\Publish\Core\Repository\Values\User\Policy` `ValueObject`. It's kind-of comparison method, so IMHO it doesn't break ValueObject pattern.

WDYT @andrerom, @pspanja?

Comment (andrerom) [inline eZ/Publish/Core/Repository/Values/User/Policy.php]: I'm not a big fan, name is wrong ;)

Also I would prefer if we could handle the logic on SPI objects which are more light before `buildDomainPolicyObject` is called.

**Comment by alongosz** [comment]
> I'm not a big fan, name is wrong ;)
> Also I would prefer if we could handle the logic on SPI objects which are more light before buildDomainPolicyObject is called.

Ok :) I've improved code in ea36e3b. For sure it looks simpler now. It turns out that with this change adjustments `\eZ\Publish\Core\Repository\Tests\Service\Mock\RepositoryTest` are not needed.

ping @andrerom, @pspanja

File: eZ/Publish/SPI/Persistence/User/Policy.php [hunk 2]
```diff
@@ -81,4 +81,25 @@ class Policy extends ValueObject
...
      */
     public $limitations;
+
+    /**
+     * Return true if at least one of the given policies overlaps this policy (has a wider scope).
+     *
+     * @param \eZ\Publish\SPI\Persistence\User\Policy[] $policies
+     * @return bool
+     */
+    public function overlaps(array $policies)
```
Comment (andrerom) [inline eZ/Publish/SPI/Persistence/User/Policy.php]: Move to be private method on `Repository` _(actually, please rebase to 6.6, as we are about to tag last tag og 6.5, so that will be PermissionResolver)_. And while at it change to something like `private function isOverlappedByWiderPolicy(spiPolicy $policy, array $policies)`

**Comment by alongosz** [comment]
Improved naming in 72015b0.

**Comment by alongosz** [comment]
Fixups:
- [`85a9b73`](https://github.com/ezsystems/ezpublish-kernel/pull/1819/commits/85a9b7302eb9ebd3f94b12d013ab668cba8a4333): Added tests testing negative case (when content should not be found), also improved naming.
- 43d2a7a: Improved readability of the method `isOverlappedByWiderRolePolicy`

// cc @andrerom @pspanja

File: eZ/Publish/Core/Repository/Repository.php [hunk 3]
```diff
@@ -454,7 +459,39 @@ public function hasAccess($module, $function, APIUserReference $user = null)
...
+     * Return true if at least one of the given policies overlaps $policy (has a wider scope).
+     *
+     * @param \eZ\Publish\SPI\Persistence\User\Policy $policy
+     * @param \eZ\Publish\SPI\Persistence\User\Policy[] $rolePolicies
+     * @return bool
+     */
+    private function isOverlappedByWiderRolePolicy(Policy $policy, array $rolePolicies)
+    {
+        foreach ($rolePolicies as $rolePolicy) {
+            // a policy can overlap other policy only if it has no limitations
```
Comment (andrerom) [inline eZ/Publish/Core/Repository/Repository.php]: this can be reflected on method as well.

Comment (alongosz) [inline eZ/Publish/Core/Repository/Repository.php]: Added to method doc in eabf437.

File: eZ/Publish/Core/Repository/Repository.php [hunk 4]
```diff
@@ -454,7 +459,40 @@ public function hasAccess($module, $function, APIUserReference $user = null)
...
+
+    /**
+     * Return true if at least one of the given policies overlaps $policy (has a wider scope).
+     * A policy can overlap other policy only if it has no limitations.
+     *
+     * @param \eZ\Publish\SPI\Persistence\User\Policy $policy
+     * @param \eZ\Publish\SPI\Persistence\User\Policy[] $rolePolicies
+     * @return bool
+     */
+    private function isOverlappedByWiderRolePolicy(Policy $policy, array $rolePolicies)
```
Comment (andrerom) [inline eZ/Publish/Core/Repository/Repository.php]: Ok, got slightly better name now: _isOverlappedByPolicyOnSameRole_

But up to @pspania now

**Comment by pspanja** [comment]
If I analyzed this correctly, this affects only search, since the problem is here:

https://github.com/ezsystems/ezpublish-kernel/blob/14698aee812c146bcb74ebf9236470eb97489c20/eZ/Publish/Core/Repository/PermissionsCriterionHandler.php#L119

If we were to add `MatchAll` criterion before `continue`, it would be another way to fix this.

@andrerom @alongosz Could you confirm this is correct?

**Comment by alongosz** [comment]
> ```php
> if ($limitations === '*' || empty($limitations)) {
>    $policyOrCriteria[] = new Criterion\MatchAll();
>    continue;
> }
> ```

@pspanja I did some tests and it looks like it's a perfect one-line solution :) 
I also tested negative case (where content should not be found despite no limitations).

When approaching this, I decided to return proper `$permissionSets` from `hasAccess`, because `hasAccess` is more widely used than `getPermissionsCriterion`. However, looking through usages of `hasAccess` in `ezpublish-kernel` repo, it seems it's enough to fix it here.

// cc @andrerom

**Comment by andrerom** [comment]
damn, ok, open a competing PR then and lets see it.

**Comment by alongosz** [comment]
New PR: #1826
